src/plugins/directoryvalue/directoryvalue_delegate.cpp

```diff
@@ -0,0 +1,524 @@
+}
+
+/**
+ * @brief This function removes the basename (last level) from the given keys.
+ *
+ * @param keys This parameter contains the keys for which this function removes the basename.
+ *
+ * @return A copy of the input, where the last level of each key was removed
+ */
+CppKeySet removeBaseName (CppKeySet const & keys)
+{
+	CppKeySet directories;
+
+	for (auto key : keys)
+	{
+		CppKey directory = key.dup ();
+		keySetBaseName (*directory, 0);
+		directories.append (directory);
+	}
+	return directories;
+}
+
+/**
+ * @brief This function converts the given array leaves to directory keys.
+ *
+ * @param arrayLeaves This parameter contains array leaves (that start with `ARRAY_VALUE_PREFIX` and have the baseName = `#0`).
+ *
+ * @return A copy of `arrayLeaves`, where key values does not start with `ARRAY_VALUE_PREFIX` any more and the baseName of each key (`#0`)
+ *         was removed
+ */
+CppKeySet convertArrayLeaves (CppKeySet const & arrayLeaves)
+{
+	CppKeySet directories = removeBaseName (arrayLeaves);
+	for (auto key : directories)
+	{
+		ELEKTRA_LOG_DEBUG ("Convert array leaf “%s”", key.getName ().c_str ());
+		if (key.getStringSize () == arrayValuePrefixSize + 1)
+		{
+			key.setBinary (0, 0);
+			ELEKTRA_LOG_DEBUG ("Set value of “%s” to NULL", key.getName ().c_str ());
+		}
+		else
+		{
+			key.setString (key.getString ().substr (arrayValuePrefixSize + 1));
+			ELEKTRA_LOG_DEBUG ("Set value of “%s” to “%s”", key.getName ().c_str (), key.getString ().c_str ());
+		}
+	}
+	return directories;
+}
+
+/**
+ * @brief This function returns a modified copy of `child`, where child is directly below `parent`.
+ *
+ * For example, if `child` has the name `user/parent/level1/level2/level3` and parent has the name `user/parent`, then the function will
+ * return a key with the name `user/parent/level1`.
+ *
+ * @pre The key `child` has to be below `parent`.
+ *
+ * @param parent This parameter specifies a parent key of `child`.
+ * @param keys This variable stores a child key of `parent`.
+ *
+ * @return A copy of `child` that is directly below `parent`
+ */
+CppKey convertToDirectChild (CppKey const & parent, CppKey const & child)
+{
+	ELEKTRA_ASSERT (child.isBelow (parent), "The key `child` is not located below `parent`");
+
+	CppKey directChild = child.dup ();
+	while (!directChild.isDirectBelow (parent))
+	{
+		keySetBaseName (*directChild, 0);
+	}
+	return directChild;
+}
+
+/**
+ * @brief This function determines if the given key is an array parent.
+ *
+ * @param parent This parameter specifies a possible array parent.
+ * @param keys This variable stores the key set of `parent`.
+ *
+ * @retval true If `parent` is the parent key of an array
+ * @retval false Otherwise
+ */
+bool isArrayParent (CppKey const & parent, CppKeySet const & keys)
+{
+	CppKeySet children = accumulate (keys.begin (), keys.end (), CppKeySet{}, [&parent](CppKeySet collected, CppKey key) {
+		if (key.isBelow (parent)) collected.append (key);
```

Comment:
Its better to have {} around here.

Reply:
Since the statement fits on a single line I like to keep it like it is.

Reply:
Ok, because it would make little sense to have the return within the if.